pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -0,0 +1,149 @@
+// other than testing.
+func NewDefaultStatefulSetControl(podControl StatefulPodControlInterface) StatefulSetControlInterface {
+	return &defaultStatefulSetControl{podControl}
+}
+
+type defaultStatefulSetControl struct {
+	podControl StatefulPodControlInterface
+}
+
+func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, pods []*v1.Pod) error {
+	if set == nil {
+		return nilParameterError
+	}
+	replicaCount := int(*set.Spec.Replicas)
+	replicas := make([]*v1.Pod, replicaCount)
+	condemned := make([]*v1.Pod, 0, len(pods))
+	ready := 0
+	unhealthy := 0
+
+	// First we partition pods into two lists, replicas will contain all Pods such that
+	// 0 <= getOrdinal(pod) < set.Spec.Replicas and condemned will contain all Pods such that
+	// set.Spec.Replicas <= getOrdinal(pod)
+	for i := range pods {
+		//count the number of running and ready replicas
+		if isRunningAndReady(pods[i]) {
+			ready++
+		}
+		if ord := getOrdinal(pods[i]); 0 <= ord && ord < replicaCount {
+			// if the ordinal of the pod is withing the range of the current number of replicas,
+			// insert it at the indirection of its ordinal
+			replicas[ord] = pods[i]
+
+		} else if ord >= replicaCount {
+			// if the ordinal is greater than the number of replicas add it to the condemned list
+			condemned = append(condemned, pods[i])
+		}
+	}
+
+	// for any empty indices in the sequence [0,set.Spec.Replicas) create a new Pod
+	for ord := 0; ord < replicaCount; ord++ {
+		if replicas[ord] == nil {
+			replicas[ord] = newStatefulSetPod(set, ord)
+		}
+		unhealthy++
+	}
+
+	// count the number of unhealthy pods
+	for i := range replicas {
+		if !isHealthy(replicas[i]) {
+			unhealthy++
+		}
+	}
+	for i := range condemned {
+		if !isHealthy(condemned[i]) {
+			unhealthy++
+		}
+	}
+
+	// sort the condemned Pods by their ordinals
+	sort.Sort(ascendingOrdinal(condemned))
+
+	// attempt to update set with the current number of replicas
+	set.Status = apps.StatefulSetStatus{Replicas: int32(ready)}
+	if err := ssc.podControl.UpdateStatefulSetStatus(set); err != nil {
+		return err
+	}
+
+	// Examine each replica with respect to its ordinal
+	for i := range replicas {
+		// delete and recreate failed pods
+		if isFailed(replicas[i]) {
+			glog.V(2).Infof("StatefulSet %s is recreating failed Pod %s", set.Name, replicas[i].Name)
+			if err := ssc.podControl.DeleteStatefulPod(set, replicas[i]); err != nil {
+				return err
+			}
+			replicas[i] = newStatefulSetPod(set, i)
+		}
+		if isCreated(replicas[i]) {
+			// if we have a Pod that has been created but is not running and ready we can not make progress
+			if !isRunningAndReady(replicas[i]) {
+				glog.V(2).Infof("StatefulSet %s is waiting for Pod %s to be Running and Ready",
+					set.Name, replicas[i].Name)
+				return nil
+			} else {
+				// enforce the StatefulSet invariants
+				if err := ssc.podControl.UpdateStatefulPod(set, replicas[i]); err != nil {
+					return err
+				}
+			}
+		} else {
```

Comment:
Invert the else so that simple conditions come first and exit, then you don't need an else